### PR TITLE
export ml options

### PR DESCRIPTION
**Pull request by songtianyi**
You may know, in China mainland, it's very slow to download content from Github. 
So I need to give go-dpi a local path to init itself.

### CODE REVIEW

**Comment by songtianyi** [comment]
There are some problems with commits, but they won't make the result bad, does they?

**Comment by glaslos** [comment]
No worries about the commit, we can squash them before merging. Can you merge master into your branch? It fixes the tests so we can see better what you are up to :)

**Comment by songtianyi** [comment]
> No worries about the commit, we can squash them before merging. Can you merge master into your branch? It fixes the tests so we can see better what you are up to :)

Done. But Travis build failed
```
Unable to find image 'godpi-example:latest' locally
docker: Error response from daemon: pull access denied for godpi-example, repository does not exist or may require 'docker login'.
```

**Comment by nikofil** [comment]
That's due to the image not being built, due to:
```
# github.com/mushorg/go-dpi/modules/ml [github.com/mushorg/go-dpi/modules/ml.test]

modules/ml/linearsvc_test.go:14:30: not enough arguments in call to NewLinearSVCModule
	have ()
	want (string, string, float32)
```

Could you make sure the tests also run with yoyr changes?

**Comment by songtianyi** [comment]
> That's due to the image not being built, due to:
> 
> ```
> # github.com/mushorg/go-dpi/modules/ml [github.com/mushorg/go-dpi/modules/ml.test]
> 
> modules/ml/linearsvc_test.go:14:30: not enough arguments in call to NewLinearSVCModule
> 	have ()
> 	want (string, string, float32)
> ```
> 
> Could you make sure the tests also run with yoyr changes?

Hi, i think it's ok now. Very sorry for the horrible commits :(
Found a better way to pass options, so i rewrite it. The minimum model
```go
package main

import "fmt"

type AnyOther struct{}

type Module interface {
	Class(*AnyOther)
}

type AModule struct {
}

func (m *AModule) Class(a *AnyOther) {
}

type Option interface {
	Apply(Module)
}

type MLOption struct {
}

func (o MLOption) Apply(m Module) {
	fmt.Println(m.(*AModule))
}

var modules = []Module{
	&AModule{},
}

func create(opts ...Option) {
	for _, opt := range opts {
		switch opt.(type) {
		case MLOption:
			opt.Apply(modules[0])
		}
		fmt.Println(opt)
	}
}
func main() {
	create()
}
```

File: godpi_example/example_app.go [hunk 1]
```diff
@@ -44,13 +44,16 @@ func main() {
 		return
 	}
 
+	//mlo := godpi.MLOption{TCPModelPath: "../2grams_tcp.model", UDPModelPath: "../2grams_udp.model", Threshold: 0.8}
+	//initErrs := godpi.Initialize(mlo)
```
Comment (glaslos) [inline godpi_example/example_app.go:48]: remove commented code

File: godpi_example/example_app.go [hunk 2]
```diff
@@ -89,6 +92,7 @@ func main() {
 		select {
 		case <-signalChannel:
 			fmt.Println("Received interrupt signal")
+			fmt.Println(protoCounts)
```
Comment (glaslos) [inline godpi_example/example_app.go:95]: this should be printed here already: https://github.com/mushorg/go-dpi/blob/77d9d9f2a2a4f6e30f4ab8e5ac8403936f5cf674/godpi_example/example_app.go#L63

File: godpi.go [hunk 3]
```diff
@@ -18,8 +19,39 @@ var moduleList = []types.Module{
 }
 var cacheExpiration = 5 * time.Minute
 
+// Options allow end users init module with custom options
+type Options interface {
+	Apply(types.Module)
+}
+
+type MLOption struct {
+	TCPModelPath, UDPModelPath string
```
Comment (glaslos) [inline godpi.go:28]: two lines for readability

File: godpi.go [hunk 4]
```diff
@@ -18,8 +19,39 @@ var moduleList = []types.Module{
...
+	Threshold                  float32
+}
+
+func (o MLOption) Apply(mod types.Module) {
+	lsm := mod.(*ml.LinearSVCModule)
```
Comment (glaslos) [inline godpi.go:33]: use a type switch statement here

File: godpi.go [hunk 5]
```diff
@@ -18,8 +19,39 @@ var moduleList = []types.Module{
...
+	}
+}
+
 // Initialize initializes the library and the selected modules.
-func Initialize() (errs []error) {
+func Initialize(opts ...Options) (errs []error) {
+	for _, opt := range opts {
+		switch opt.(type) {
+		case MLOption:
+			opt.Apply(moduleList[2])
```
Comment (glaslos) [inline godpi.go:50]: can you improve this? looks error prone to me.

Comment (songtianyi) [reply godpi.go]:
```go
 lsm, ok := mod.(*ml.LinearSVCModule)
        if !ok {
                return
        }
```
type convert check, it that ok for you?

File: godpi.go (hunk 4 above)
Comment (songtianyi) [reply godpi.go]:
```go
 lsm, ok := mod.(*ml.LinearSVCModule)
        if !ok {
                return
        }
```
type convert check, it that ok for you?

File: godpi.go (hunk 5 above)
Comment (songtianyi) [reply godpi.go]: Yes. use index my cause out of range panic when some guys modified the default module list or call SetModules.
The best way i can work out is by using reflect, any suggestions?

Comment (nikofil) [reply godpi.go]: Can't we simply call Apply on all options, with all modules? Then the option itself does what it wants with it, checking if it's been called on the appropriate module (as you currently do).

File: godpi.go (hunk 4 above)
Comment (nikofil) [reply godpi.go]: That should be fine.

Comment (nikofil) [reply godpi.go]: Maybe we could move this option to a new file however in the same package (options.go?) to keep the module-specific logic outside godpi.go.

Comment (songtianyi) [reply godpi.go]: Sure

File: godpi.go (hunk 5 above)
Comment (songtianyi) [reply godpi.go]: Done.

**Comment by nikofil** [comment]
LGTM :+1:

**Review by glaslos — APPROVED** [review]

**Comment by glaslos** [comment]
Sorry, totally missed this 🤦 Thank you for the contribution!

**Pull request merged**